util/template/client-class

```diff
@@ -60,19 +67,37 @@ class Client
     /**
      * Client constructor
      *
-     * @param Transport                   $transport
-     * @param callable                    $endpoint
+     * @param Transport $transport
+     * @param callable|EndpointFactoryInterface $endpointFactory
      * @param NamespaceBuilderInterface[] $registeredNamespaces
      */
-    public function __construct(Transport $transport, callable $endpoint, array $registeredNamespaces)
+    public function __construct(Transport $transport, callable|EndpointFactoryInterface $endpointFactory, array $registeredNamespaces)
     {
         $this->transport = $transport;
-        $this->endpoints = $endpoint;
+        if (is_callable($endpointFactory)) {
+            @trigger_error('Passing a callable to AbstractNamespace is deprecated in 2.3.2 and will be removed in 3.0.0. Pass an instance of \OpenSearch\EndpointFactoryInterface instead.', E_USER_DEPRECATED);
+            $this->endpoints = $endpointFactory;
```

Comment:
If you want to provide full BC for child classes using the protected `endpoints` property, you actually need a `else` clause assigning a function that would use the EndpointFactory. Otherwise, you are still breaking BC by changing the type of the property to `callable|null` (as it will be `null` when not assigning a callable as it is not a typed property).

Btw, this compat callable could be a good place to trigger a deprecation saying that the `endpointFactory` property should be used instead of the `endpoints` callable (as the current deprecation triggered in `__get` won't actually be triggered)